okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java

```diff
@@ -183,18 +191,72 @@ public OkHttpClientStream newStream(MethodDescriptor<?, ?> method,
     List<Header> requestHeaders =
         Headers.createRequestHeaders(headers, defaultPath, defaultAuthority);
 
+    SettableFuture<Void> pendingFuture = null;
     synchronized (lock) {
       if (goAway) {
-        throw new IllegalStateException("Transport not running", goAwayStatus.asRuntimeException());
+        clientStream.transportReportStatus(goAwayStatus, true, new Metadata.Trailers());
+      } else if (!goAway && streams.size() >= maxConcurrentStreams) {
+        pendingFuture = SettableFuture.create();
+        pendingStreams.add(new PendingStream(clientStream, pendingFuture, requestHeaders));
+      } else {
+        startStream(clientStream, requestHeaders);
+      }
+    }
+
+    if (pendingFuture != null) {
+      try {
+        pendingFuture.get();
+      } catch (InterruptedException e) {
+        // Restore the interrupt.
+        Thread.currentThread().interrupt();
+        clientStream.cancel();
+        throw new RuntimeException(e);
+      } catch (ExecutionException e) {
+        clientStream.cancel();
+        throw new RuntimeException(e.getCause() != null ? e.getCause() : e);
       }
-      
-      assignStreamId(clientStream);
-      frameWriter.synStream(false, false, clientStream.id(), 0, requestHeaders);
     }
 
     return clientStream;
   }
 
+  @GuardedBy("lock")
+  private void startStream(OkHttpClientStream stream, List<Header> requestHeaders) {
+    Preconditions.checkState(stream.id() == null, "StreamId already assigned");
+    stream.id(nextStreamId);
+    streams.put(stream.id(), stream);
+    if (nextStreamId >= Integer.MAX_VALUE - 2) {
+      onGoAway(Integer.MAX_VALUE, Status.INTERNAL.withDescription("Stream ids exhausted"));
```

Comment:
It is now obvious that the previous behavior was wrong here, as we send synStream with a negative id(). It looks like we can just return after the onGoAway.